app/uk/gov/hmrc/cgtpropertydisposals/actors/TaxEnrolmentRetryManager.scala

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright 2019 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.cgtpropertydisposals.actors
+
+import java.util.concurrent.TimeUnit
+
+import akka.actor.{Actor, Props, Scheduler}
+import akka.pattern.pipe
+import com.typesafe.config.{Config, ConfigFactory}
+import net.ceedubs.ficus.Ficus._
+import uk.gov.hmrc.cgtpropertydisposals.actors.TaxEnrolmentRetryManager.{AttemptTaxEnrolmentAllocationToGroup, CallTaxEnrolmentService, RetryTaxEnrolmentRequest}
+import uk.gov.hmrc.cgtpropertydisposals.connectors.TaxEnrolmentConnector
+import uk.gov.hmrc.cgtpropertydisposals.models.TaxEnrolmentRequest
+import uk.gov.hmrc.cgtpropertydisposals.repositories.TaxEnrolmentRetryRepository
+import uk.gov.hmrc.cgtpropertydisposals.util.Logging
+import uk.gov.hmrc.http.HeaderCarrier
+
+import scala.concurrent.duration.{FiniteDuration, _}
+
+class TaxEnrolmentRetryManager(
+  taxEnrolmentConnector: TaxEnrolmentConnector,
+  taxEnrolmentRetryRepository: TaxEnrolmentRetryRepository,
+  scheduler: Scheduler
+) extends Actor
+    with Logging {
+  import context.dispatcher
+  implicit val hc: HeaderCarrier = HeaderCarrier()
+
+  val config: Config = ConfigFactory.load()
+
+  val minBackoff: FiniteDuration = config.as[FiniteDuration]("tax-enrolment-retry.min-backoff")
+  val maxBackoff: FiniteDuration = config.as[FiniteDuration]("tax-enrolment-retry.max-backoff")
+  val numberOfRetriesBeforeDoublingInitialWait: Int =
+    config.as[Int]("tax-enrolment-retry.number-of-retries-until-initial-wait-doubles")
+  val maxAllowableElapsedTime: FiniteDuration = config.as[FiniteDuration]("tax-enrolment-retry.max-elapsed-time")
+  val maxWaitTimeForSubmittingRecoveredEnrolmentRequests: Int =
+    config.as[Int]("tax-enrolment-retry.max-time-to-wait-after-recovery")
+
+  def is5xx(status: Int): Boolean = status >= 500 && status < 600
+
+  override def preStart(): Unit = {
+    super.preStart()
+    taxEnrolmentRetryRepository.getAllNonFailedEnrolmentRequests().value.map {
+      case Left(error) => logger.warn(s"Could not recover tax enrolment retry requests: $error")
```

Comment:
if this fails, it seems like you should retry this again so that this actor can get on with retrying what it needs to